src/data/articles/scala-3-the-evolution-of-a-macro.mdx

```diff
@@ -0,0 +1,1829 @@
+
+[^noPun]: No pun here, none at all.
+
+```scala wrap=false
+def isGiven(using Quotes)(param: quotes.reflect.Symbol) =
+  import quotes.reflect.*
+
+  val flags = param.flags // 1
+
+  flags.is(Flags.Given) || flags.is(Flags.Implicit) // 2
+```
+
+`isGiven` checks whether a `Symbol` that corresponds to a method argument is a given parameter. We do this by extracting the symbol's `Flags` (1), these contain lots of useful metadata about symbols. We then verify that the `Symbol` is neither `given` nor `implicit` (for legacy reasons) (2).
+
+```scala wrap=false
+def isByName(using Quotes)(param: quotes.reflect.Symbol) =
+  import quotes.reflect.*
+
+  param.info match // 1
+    case ByNameType(_) => true // 2
+    case _ => false
+
+```
+
+Similarly, `isByName` checks whether a method argument is a by-name argument. We do this by getting the `info` field of the parameter (1), which contains type information about the `Symbol`. By matching with the custom extractor `ByNameType` (2) on the `info` value, we can decide whether this is a by-name argument or not.
+
+```scala wrap=false
+def isUntraceable(using Quotes)(param: quotes.reflect.Symbol) =
+  import quotes.reflect.*
+
+  // 1
+  param.info =:= TypeRepr.of[UntraceableData]
+```
+
+For reasons that will become clear later on, we have a helper that checks whether a symbol is of the `UntraceableData` type. We do this again by using the `info` field, and we type-compare it to the `TypeRepr` of `UntraceableData` (1).
+
+With these helpers in hand, we are ready to tackle `trace5`. Brace yourselves:
+
+```scala wrap=false
+object DataFromSignature:
+  inline def trace5[A](level: TraceLevel) // 1
+                      (inline body: A)
+                      (using tracer: Tracer): A =
+    ${ trace5Impl('level, 'body, 'tracer) }
+
+  def trace5Impl[A: Type](level: Expr[TraceLevel],
+                          body: Expr[A],
+                          tracer: Expr[Tracer])
+                         (using Quotes): Expr[A] =
+    import quotes.reflect.*
+
+    val enclosingMethod = getEnclosingMethodSymbol // 2
+
+    val allParams = enclosingMethod.paramSymss // 3
+      .flatten // 4
+
+    val onlyTermParams = allParams.filter(_.isTerm) // 5
+
+    val params = onlyTermParams
+      .filterNot(isGiven) // 6
+      .filterNot(isByName) // 7
+      .filterNot(isUntraceable) // 8
+
+    def toField(param: Symbol) = // 9
+      val fieldName = Expr(param.name) // 10
+      val fieldValue = Ref(param).asExpr // 11
+
+      '{ $fieldName -> $fieldValue.toString } // 12
+
+    val fieldsList = params.map(toField) // 13
+    val fields = Expr.ofSeq(fieldsList) // 14
+
+    val name = Expr(enclosingMethod.name) // 15
+
+    '{ WithInlining.trace1($level, $name, $fields*)($body)(using $tracer) } // 16
+```
+
+Step by step:
+- `trace5` now only needs the `TracingLevel` and `Tracer` to be able to function (1), no metadata necessary
```

Comment:
might have been easier to use a numbered list, especially for long examples like this - I don't mind the current structure though

Reply:
I was considering that as well. But in some of the lists not every bullet in the list refers to a numbered code line, in which case a discrepancy between the list numbering and the code numbering might lead to confusion. If it's okay, I'll keep it as it is, thinking about restructuring the lists to have a 1-1 correspondence with the code numbering makes my brain hurt...

Reply:
I wasn't referring to using the line numbers in the code, but rather numbered lists in markdown:
1. `trace5` now only needs...
2. ...
3. ...

Again, not a big deal. Will leave this as is.

Reply:
That's indeed what I meant, it's not always the case that bullets correspond to a number in code. So the automatic list numbering might diverge from the actual numbers in code.